rustls/src/crypto/mod.rs

```diff
@@ -186,6 +187,69 @@ pub struct CryptoProvider {
     pub key_provider: &'static dyn KeyProvider,
 }
 
+impl CryptoProvider {
+    fn supported_kx_algos(&self) -> Vec<KeyExchangeAlgorithm> {
+        let mut res = Vec::with_capacity(ALL_KEY_EXCHANGE_ALGORITHMS.len());
+        for kx in self
+            .kx_groups
+            .iter()
+            .filter_map(|kx| kx.name().key_exchange_algorithm())
+        {
+            if !res.contains(&kx) {
+                res.push(kx);
+            }
+            if res.len() == ALL_KEY_EXCHANGE_ALGORITHMS.len() {
+                break;
+            }
+        }
+        res
+    }
+
+    pub(crate) fn supported_kx_group_names(&self) -> Vec<NamedGroup> {
+        self.kx_groups
+            .iter()
+            .map(|skxg| skxg.name())
+            .collect::<Vec<_>>()
+    }
+
+    pub(crate) fn verify_cipher_suites_have_matching_kx(&self) -> Result<(), Error> {
+        let kx_algos = self.supported_kx_algos();
+
+        for cs in self.cipher_suites.iter() {
+            let cs_kx = cs.key_exchange_algorithms();
+            if !cs_kx
+                .iter()
+                .any(|kx| kx_algos.contains(kx))
+            {
+                let suite_name = cs.common().suite;
+                return Err(Error::General(alloc::format!(
+                    "Ciphersuite {suite_name:?} requires {cs_kx:?} key exchange, but no {cs_kx:?}-compatible \
+                    key exchange groups were present in `CryptoProvider`'s `kx_groups` field",
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    /// Removes cipher suites from this `CryptoProvider` that have no matching `SupportedKxGroup`s in `kx_groups`
+    pub fn remove_cipher_suites_without_matching_kx(&mut self) {
```

Comment:
I would conceptually expect `CryptoProvider` values to be immutable. Why do we need this method (and its caller, below)?

Reply:
With two different kx algorithms, we may get into a situation where the `CryptoProvider` offers a cipher suite but does not have a kx group corresponding to the cipher suite. 

For example, the provider could offer `TLS_DHE_RSA_WITH_AES_128_GCM_SHA256` but have no FFDHE groups in its `kx_groups` field. If such a provider is configured for the client, and the server picks `TLS_DHE_RSA_WITH_AES_128_GCM_SHA256`, the handshake cannot succeed. The purpose of this method is to prevent such cases (and be convenient for users). 

There is a related function, `verify_cipher_suites_have_matching_kx` that enforces this invariant when configuring a provider.

Reply:
I think we should just rely on verifying the invariant when configuring a provider and not have any methods that mutate an existing `CryptoProvider`. We should have documentation on the `CryptoProvider` fields that mention any invariants that must hold between separate fields.